File: rest-service/src/main/java/mb/lib/query/model/BlastJob.java
package mb.lib.query.model;

import mb.lib.blast.model.BlastQuery;
import mb.lib.model.HashID;
import mb.lib.util.JSON;
import mb.lib.util.MD5;
import org.veupathdb.lib.blast.BlastQueryConfig;
import org.veupathdb.lib.blast.BlastTool;

public class BlastJob
{
  private BlastQueryConfig config;
  private BlastQuery       query;
  private String           site;
  private String           description;
  private long             userID;
  private long             maxDLSize;
  private HashID           parentID;
  private JobTarget[]      targets;

  public String getSite() {
    return site;
  }

  public BlastJob setSite(String site) {
    this.site = site;
    return this;
  }

  public BlastQuery getQuery() {
    return query;
  }

  public BlastJob setQuery(BlastQuery query) {
    this.query = query;
    return this;
  }

  public String getDescription() {
    return description;
  }

  public BlastJob setDescription(String description) {
    this.description = description;
    return this;
  }

  public long getUserID() {
    return userID;
  }

  public BlastJob setUserID(long userID) {
    this.userID = userID;
    return this;
  }

  public long getMaxDLSize() {
    return maxDLSize;
  }

  public BlastJob setMaxDLSize(long maxDLSize) {
    this.maxDLSize = maxDLSize;
    return this;
  }

  public boolean isPrimary() {
    return parentID == null;
  }

  public HashID getParentID() {
    return parentID;
  }

  public BlastJob setParentID(HashID parentID) {
    this.parentID = parentID;
    return this;
  }

  public JobTarget[] getTargets() {
    return targets;
  }

  public BlastJob setTargets(JobTarget[] targets) {
    this.targets = targets;
    return this;
  }

  public BlastQueryConfig getConfig() {
    return config;
  }

  public BlastJob setConfig(BlastQueryConfig config) {
    this.config = config;
    return this;
  }

  public HashID digest(String query) {
    try {
      return new HashID(MD5.hash(JSON.stringify(new HashWrapper(site, query, config))));
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  static class HashWrapper
  {
    private final BlastTool        tool;
    private final String           site;
    private final String           dbFiles;
    private final String           query;
    private final BlastQueryConfig config;

    public HashWrapper(String site, String query, BlastQueryConfig config) {
      this.tool    = config.getTool();
      this.site    = site;
      this.query   = query;
      this.dbFiles = config.getDBFile();
      this.config  = config;
    }

    public BlastTool getTool() {
      return tool;
    }

    public String getSite() {
      return site;
    }

    public String getDbFiles() {
      return dbFiles;
    }

    public String getQuery() {
      return query;
    }

    public BlastQueryConfig getConfig() {
      return config;
    }
  }
}

File: rest-service/src/main/kotlin/mb/lib/query/model/BlastJob.kt
package mb.lib.query.model

import mb.lib.blast.model.BlastQuery
import mb.lib.util.jsonStringify
import mb.lib.util.md5
import mb.lib.util.toHashable
import org.veupathdb.lib.blast.BlastQueryConfig
import org.veupathdb.lib.hash_id.HashID

@Suppress("ArrayInDataClass")
data class BlastJob(
  val config:      BlastQueryConfig,
  val query:       BlastQuery,
  val site:        String,
  val description: String? = null,
  val userID:      Long,
  val maxDLSize:   Long,
  val targets:     Array<JobTarget>,
) {
  var parentID:    HashID? = null

  var isPrimary:   Boolean = true

  constructor(
    config: BlastQueryConfig,
    query: BlastQuery,
    site: String,
    description: String? = null,
    userID: Long,
    maxDLSize: Long,
    targets: Array<JobTarget>,
    parentID: HashID
  ) : this(config, query, site, description, userID, maxDLSize, targets) {
    this.parentID = parentID
    this.isPrimary = false
  }

  fun digest(query: String): HashID {
    try {
      return HashID.ofMD5(HashWrapper(site, query, config).jsonStringify())
    } catch (e: Exception) {
      throw  RuntimeException(e)
    }
  }
}

class HashWrapper(
  val site:   String,
  val query:  String,
  rawConfig: BlastQueryConfig
) {
  val tool    = rawConfig.tool
  val dbFiles = rawConfig.dbFile
  val config  = rawConfig.toHashable()
}
